test/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter_integration_test.cc

```diff
@@ -0,0 +1,129 @@
+  }
+
+  EXPECT_EQ(num_forwarded, forwarded_count);
+}
+
+IntegrationStreamDecoderPtr AdaptiveConcurrencyIntegrationTest::respondToRequest() {
+  auto response = std::move(response_q_.front());
+  response_q_.pop();
+  response->waitForEndStream();
+  EXPECT_TRUE(response->complete());
+  return response;
+}
+
+uint32_t
+AdaptiveConcurrencyIntegrationTest::inflateConcurrencyLimit(const uint64_t limit_lower_bound) {
+  // Send requests until the gauge exists.
+  while (!test_server_->gauge(kConcurrencyLimitGaugeName)) {
+    sendRequests(1);
+    respondToAllRequests(1);
+  }
+
+  while (test_server_->gauge(kConcurrencyLimitGaugeName)->value() < limit_lower_bound) {
+    const auto min_rtt = test_server_->gauge(kMinRTTGaugeName)->value();
+    sendRequests(1, min_rtt / 2);
+    respondToAllRequests(1);
+  }
+  return test_server_->gauge(kConcurrencyLimitGaugeName)->value();
+}
+
+void AdaptiveConcurrencyIntegrationTest::deflateConcurrencyLimit(const uint64_t limit_upper_bound) {
+  ASSERT(limit_upper_bound > 1);
+  // Send requests until the gauge exists.
+  while (!test_server_->gauge(kConcurrencyLimitGaugeName)) {
+    sendRequests(1);
+    respondToAllRequests(1);
+  }
+
+  // We cannot break when the concurrency limit is 1, because this implies we're in a minRTT
+  // recalculation window. This is not a decrease in the concurrency limit due to latency samples.
+  while (test_server_->gauge(kConcurrencyLimitGaugeName)->value() != 1 &&
+         test_server_->gauge(kConcurrencyLimitGaugeName)->value() >= limit_upper_bound) {
+    const auto min_rtt = test_server_->gauge(kMinRTTGaugeName)->value();
+    sendRequests(1, min_rtt * 2);
+    respondToAllRequests(1);
+  }
+}
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, AdaptiveConcurrencyIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+/**
+ * Test a single request returns successfully.
+ */
+TEST_P(AdaptiveConcurrencyIntegrationTest, TestConcurrency1) {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  sendRequests(2);
+  auto response = respondToRequest();
+  verifyResponseForwarded(std::move(response));
+  response = respondToRequest();
+  verifyResponseBlocked(std::move(response));
+}
+
+/**
+ * Test many requests, where only a single request returns 200 during the minRTT window.
+ */
+TEST_P(AdaptiveConcurrencyIntegrationTest, TestManyConcurrency1) {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  sendRequests(10);
+  respondToAllRequests(1);
+  test_server_->waitForCounterGe(kRequestBlockCounterName, 9);
+}
+
+/**
+ * Test the ability to increase/decrease the concurrency limit with request latencies based on the
+ * minRTT value.
+ */
+TEST_P(AdaptiveConcurrencyIntegrationTest, TestConcurrencyLimitMovement) {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Cause the concurrency limit to oscillate.
+  for (int ii = 0; ii < 3; ++ii) {
```

Comment:
nit: we don't use `ii` anywhere else, please replace with `i` or `index` or something else. Same in other code I didn't notice this in.